samples/SocialSample/Startup.cs

```diff
@@ -228,6 +231,99 @@ public void Configure(IApplicationBuilder app)
+                        // The cookie middleware will handle this and redirect to /login
+                        await context.ChallengeAsync();
+
+                        // This is what [Authorize(ActiveAuthenticationSchemes = MicrosoftAccountDefaults.AuthenticationScheme)] calls
+                        // await context.ChallengeAsync(MicrosoftAccountDefaults.AuthenticationScheme);
+
+                        return;
+                    }
+
+                    var refreshToken = authProperties.GetTokenValue("refresh_token");
+
+                    if (string.IsNullOrEmpty(refreshToken))
+                    {
+                        context.Response.ContentType = "text/html";
+                        await context.Response.WriteAsync("<html><body>");
+                        await context.Response.WriteAsync("No refresh_token is available.<br>");
+                        await context.Response.WriteAsync("<a href=\"/\">Home</a>");
+                        await context.Response.WriteAsync("</body></html>");
+                    }
+
+                    if (string.Equals(GoogleDefaults.AuthenticationScheme, user.Identities.First().AuthenticationType))
+                    {
+                        var handler = (AuthenticationHandler<GoogleOptions>)await authSchemes.GetHandlerAsync(context, GoogleDefaults.AuthenticationScheme);
+
+                        var pairs = new Dictionary<string, string>()
+                        {
+                            { "client_id", handler.Options.ClientId },
+                            { "client_secret", handler.Options.ClientSecret },
+                            { "grant_type", "refresh_token" },
+                            { "refresh_token", refreshToken }
+                        };
+                        var content = new FormUrlEncodedContent(pairs);
+                        var response = await handler.Options.Backchannel.PostAsync(GoogleDefaults.TokenEndpoint, content, context.RequestAborted);
+                        response.EnsureSuccessStatusCode();
+
+                        var payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+
+                        // Persist the new acess token
+                        authProperties.UpdateTokenValue("access_token", payload.Value<string>("access_token"));
+                        if (int.TryParse(payload.Value<string>("expires_in"), NumberStyles.Integer, CultureInfo.InvariantCulture, out var seconds))
+                        {
+                            var expiresAt = DateTimeOffset.UtcNow + TimeSpan.FromSeconds(seconds);
+                            authProperties.UpdateTokenValue("expires_at", expiresAt.ToString("o", CultureInfo.InvariantCulture));
+                        }
+                        await context.SignInAsync(user, authProperties);
+
+                        context.Response.ContentType = "text/html";
+                        await context.Response.WriteAsync("<html><body>");
```

Comment:
nit: so many context.Response, store it in a var?  Think of all the poor chars that will be saved!